### PR TITLE
[IMP]solve the errors and improvements

### PR DESCRIPTION
**Pull request by vimalpatelserpentcs**

### CODE REVIEW

**Comment by JayVora-SerpentCS** [comment]
travis fix needed.

File: assignment/models/homework.py [hunk 1]
```diff
@@ -81,6 +81,14 @@ def done_assignments(self):
         self.state = 'done'
         return True
 
+    @api.multi
+    def unlink(self):
+        for rec in self:
+            if rec.state != 'draft':
+                raise ValidationError(_('''You can delete record in
+                                        draft state only.'''))
+            return super(SchoolTeacherAssignment, self).unlink()
```
Comment (JayVora-SerpentCS) [inline assignment/models/homework.py]: shouldn't you shift the life left?

File: assignment/models/homework.py [hunk 2]
```diff
@@ -143,3 +157,11 @@ def reassign_assignment(self):
...
         self.state = 'active'
         return True
+
+    @api.multi
+    def unlink(self):
+        for rec in self:
+            if rec.state != 'draft':
+                raise ValidationError(_('''You can delete record
+                                        in draft state only.'''))
+            return super(SchoolStudentAssignment, self).unlink()
```
Comment (JayVora-SerpentCS) [inline assignment/models/homework.py]: shift left. it should not be inside loop

File: exam/models/exam.py [hunk 3]
```diff
@@ -31,30 +31,59 @@ class ExtendedTimeTable(models.Model):
...
-        return True
+            exam_search = exam.search([('state', '=', 'running')])
+            for data in exam_search:
+                schedule_line_search = schedule_line.search([('exam_id', '=',
+                                                              data.id),
+                                                             ('timetable_id',
+                                                              '=', rec.id)])
+                if schedule_line_search:
+                    raise ValidationError(_('''You cannot delete schedule
+                                           of exam in running state'''))
```
Comment (JayVora-SerpentCS) [inline exam/models/exam.py]: Add an exclamation at the end.

File: exam/models/exam.py [hunk 4]
```diff
@@ -31,30 +31,59 @@ class ExtendedTimeTable(models.Model):
...
+        return super(ExtendedTimeTable, self).unlink()
 
     timetable_type = fields.Selection(selection_add=[('exam', 'Exam')],
                                       string='Time Table Type', required=True,
                                       inivisible=False)
     exam_timetable_line_ids = fields.One2many('time.table.line', 'table_id',
                                               'TimeTable')
     exam_id = fields.Many2one('exam.exam', 'Exam')
-    users = fields.Many2many('res.users', string="Users",
-                             compute="_compute_users", store=True)
```
Comment (JayVora-SerpentCS) [inline exam/models/exam.py:50]: Explain the purpose of removing the field.

File: exam/models/exam.py [hunk 5]
```diff
@@ -31,30 +31,59 @@ class ExtendedTimeTable(models.Model):
...
+                           if (rec.day_of_week == rec_check.day_of_week and
+                               rec.start_time == rec_check.start_time and
+                               rec.end_time == rec_check.end_time and
+                               rec.teacher_id.id == rec.teacher_id.id and
+                               rec.exm_date == rec.exm_date)]
+                if len(records) > 1:
+                    raise ValidationError(_('''You cannot set exam
+                                           at same time %s
+                                           at same day %s for teacher %s''') %
+                                          (rec.start_time, rec.day_of_week,
```
Comment (JayVora-SerpentCS) [inline exam/models/exam.py]: Add exclamation

File: exam/models/exam.py [hunk 6]
```diff
@@ -31,30 +31,59 @@ class ExtendedTimeTable(models.Model):
...
+                                           rec.teacher_id.name))
 
 
 class ExtendedTimeTableLine(models.Model):
     _inherit = 'time.table.line'
 
     exm_date = fields.Date('Exam Date')
     day_of_week = fields.Char('Week Day')
+    class_room_id = fields.Many2one('class.room', 'Classroom')
```
Comment (JayVora-SerpentCS) [inline exam/models/exam.py:85]: Should not it be mandatory at view level atleast?

File: exam/models/exam.py [hunk 7]
```diff
@@ -81,6 +110,62 @@ def _check_date(self):
...
+                if rec.table_id.timetable_type == 'exam':
+                    if not rec.teacher_id:
+                        raise ValidationError(_('''Enter Supervisior'''))
+
+    @api.constrains('start_time', 'end_time')
+    def check_time(self):
+        for rec in self:
+            if rec.start_time >= rec.end_time:
+                raise ValidationError(_('''Start time should be
+                                        less than end time'''))
```
Comment (JayVora-SerpentCS) [inline exam/models/exam.py]: Correct the error as : Configure the start time as earlier than end time.

File: exam/models/exam.py [hunk 8]
```diff
@@ -271,6 +356,33 @@ class AdditionalExam(models.Model):
     weightage = fields.Char("WEIGHTAGE")
     create_date = fields.Date("Created Date", help="Exam Created Date")
     write_date = fields.Date("Updated date", help="Exam Updated Date")
+    color_name = fields.Integer("Color index of creator",
+                                compute='_compute_color_name', store=False)
+
+    @api.multi
+    def _compute_color_name(self):
+        for rec in self:
+            rec.color_name = rec.subject_id.id
```
Comment (JayVora-SerpentCS) [inline exam/models/exam.py]: make sure compute functions are predefined. Place them before the field declaration.

File: exam/models/exam.py [hunk 9]
```diff
@@ -271,6 +356,33 @@ class AdditionalExam(models.Model):
...
+        if self.minimum_marks > self.maximum_marks:
+            raise ValidationError(_('''Maximum marks should be greater
+                                    than minimum marks'''))
+
+    @api.model
+    def create(self, vals):
+        curr_dt = datetime.now()
+        new_dt = datetime.strftime(curr_dt, '%m/%d/%Y')
+        vals.update({'create_date': new_dt})
+        return super(AdditionalExam, self).create(vals)
```
Comment (JayVora-SerpentCS) [inline exam/models/exam.py:369]: Cannot we set them in default? And why we set the field which is preset? the field is among 5 preset fields, isnt it?

File: exam/models/exam.py [hunk 10]
```diff
@@ -354,6 +466,22 @@ def write(self, vals):
                          })
         return super(ExamResult, self).write(vals)
 
+    @api.multi
+    def unlink(self):
+        for rec in self:
+            if rec.state != 'draft':
+                raise ValidationError(_('''You can delete record in
+                                        draft state only.'''))
+            return super(ExamResult, self).unlink()
```
Comment (JayVora-SerpentCS) [inline exam/models/exam.py]: move left.

File: exam/report/result_info.py [hunk 11]
```diff
@@ -52,6 +57,13 @@ def get_exam_data(self, result_id, student):
     @api.model
     def render_html(self, docids, data=None):
         docs = self.env['student.student'].browse(docids)
+        for rec in docs:
+            res_search = self.env['exam.result'].search([('student_id', '=',
+                                                          rec.id)])
+            if not res_search or rec.state == 'draft':
+                raise ValidationError(_('''You can not print report for
```
Comment (JayVora-SerpentCS) [inline exam/report/result_info.py]: You cannot, not 'You can not'

File: exam/security/exam_security.xml [hunk 12]
```diff
@@ -22,11 +22,11 @@
...
         <record id="timetable_teacher_record_rule_exam" model="ir.rule">
-	        <field name="name">Teacher Timetable Record Rule</field>
-	        <field name="model_id" ref="model_time_table"/>
-	        <field name="global" eval="True"/>
-	        <field name="domain_force">[('user_ids','=',user.id),('timetable_type','=','exam')]</field>
-	        <field name="groups" eval="[(4, ref('school.group_school_teacher'))]"/>
+            <field name="name">Teacher Timetable Record Rule</field>
+            <field name="model_id" ref="model_time_table"/>
+            <field name="global" eval="True"/>
+            <field name="domain_force">[('user_ids','=',user.id)]</field>
```
Comment (JayVora-SerpentCS) [inline exam/security/exam_security.xml:38]: does the field user_ids exist?

File: library/models/library.py [hunk 13]
```diff
@@ -83,16 +84,65 @@ def write(self, vals):
...
+    def librarycard_expire(self):
+        '''Schedular to change in librarycard state when end date
+            is over'''
+        current_date = datetime.now()
+        new_date = datetime.strftime(current_date, '%m/%d/%Y')
+        lib_card = self.env['library.card']
+        lib_card_search = lib_card.search([('end_date', '<', new_date)])
+        if lib_card_search:
+            for rec in lib_card_search:
+                rec.state = 'expire'
```
Comment (JayVora-SerpentCS) [inline library/models/library.py]: Do you think loop is needed?

File: library/models/library.py [hunk 14]
```diff
@@ -205,24 +255,22 @@ def _check_issue_book_limit(self):
                     else:
                         # Check the book issue limit on card if it is over it
                         # give warning
-                        raise UserError(_('Book issue limit is over on this\
-                        card'))
+                        raise ValidationError(_('''Book issue limit is over
+                                                on this  card'''))
```
Comment (JayVora-SerpentCS) [inline library/models/library.py]: correct spacing between 'this' and 'card'. add exclamation.

File: library/models/library.py [hunk 15]
```diff
@@ -296,17 +359,31 @@ def create(self, vals):
...
+                              ].search([('name', '=',
+                                         vals.get('name')),
+                                        ('card_id', '=',
+                                         vals.get('card_id')),
+                                        ('state', 'not in', ['draft', 'cancel',
+                                                             'return', 'paid'
+                                                             ])])
+        if book_issue:
+            raise ValidationError(_('''You cannot issue same book on
+                                    same card more than one time'''))
```
Comment (JayVora-SerpentCS) [inline library/models/library.py]: more than once!

File: library/models/library.py [hunk 16]
```diff
@@ -345,10 +423,20 @@ def issue_book(self):
...
+        if self.issue_code == 'New':
+            self.issue_code = self.env['ir.sequence'
+                                       ].next_by_code('library.book.issue'
+                                                      ) or _('New')
         for rec in self:
             if rec.name and rec.name.availability == 'notavailable':
-                raise ValidationError(_('This Book is not available!'
-                                        '\nPlease try after sometime !'))
+                raise ValidationError(_('''This Book is not available
+                                        Please try after sometime!'''))
```
Comment (JayVora-SerpentCS) [inline library/models/library.py]: newline was needed!

File: school_event/models/event.py [hunk 17]
```diff
@@ -203,6 +229,46 @@ def regi_cancel(self):
...
+                                        event which is in running or
+                                        closed state'''))
+
+    @api.multi
+    def unlink(self):
+        for rec in self:
+            if rec.state != 'draft':
+                raise UserError(_('''You can delete record
+                                 in draft state only.'''))
+            return super(SchoolEventRegistration, self).unlink()
```
Comment (JayVora-SerpentCS) [inline school_event/models/event.py]: loop should not return.

File: exam/models/exam.py [hunk 18]
```diff
@@ -271,6 +356,33 @@ class AdditionalExam(models.Model):
...
+        new_dt = datetime.strftime(curr_dt, '%m/%d/%Y')
+        vals.update({'create_date': new_dt})
+        return super(AdditionalExam, self).create(vals)
+
+    @api.multi
+    def write(self, vals):
+        curr_dt = datetime.now()
+        new_dt = datetime.strftime(curr_dt, '%m/%d/%Y')
+        vals.update({'write_date': new_dt})
+        return super(AdditionalExam, self).write(vals)
```
Comment (JayVora-SerpentCS) [inline exam/models/exam.py:376]: Do we need?

File: assignment/models/homework.py (hunk 1 above)
Comment (vimalpatelserpentcs) [reply assignment/models/homework.py]: done

File: exam/models/exam.py (hunk 7 above)
Comment (vimalpatelserpentcs) [reply exam/models/exam.py]: done

File: exam/report/result_info.py (hunk 11 above)
Comment (vimalpatelserpentcs) [reply exam/report/result_info.py]: done

File: library/models/library.py (hunk 15 above)
Comment (vimalpatelserpentcs) [reply library/models/library.py]: done

File: exam/models/exam.py (hunk 3 above)
Comment (vimalpatelserpentcs) [reply exam/models/exam.py]: done

File: assignment/models/homework.py (hunk 2 above)
Comment (vimalpatelserpentcs) [reply assignment/models/homework.py]: done

File: exam/models/exam.py (hunk 6 above)
Comment (vimalpatelserpentcs) [reply exam/models/exam.py:85]: done

File: exam/models/exam.py (hunk 5 above)
Comment (vimalpatelserpentcs) [reply exam/models/exam.py]: done

File: exam/models/exam.py (hunk 10 above)
Comment (vimalpatelserpentcs) [reply exam/models/exam.py]: done

File: library/models/library.py (hunk 14 above)
Comment (vimalpatelserpentcs) [reply library/models/library.py]: done

File: school_event/models/event.py (hunk 17 above)
Comment (vimalpatelserpentcs) [reply school_event/models/event.py]: done

File: exam/security/exam_security.xml (hunk 12 above)
Comment (vimalpatelserpentcs) [reply exam/security/exam_security.xml:38]: Yes, user_ids field exist in model time.table and we have inherited time.table in exam module.

File: exam/models/exam.py (hunk 4 above)
Comment (vimalpatelserpentcs) [reply exam/models/exam.py:50]: We have inherited model time.table in exam module and user_ids field is defined in the model time.table.So it is removed from exam.py file

**Comment by JayVora-SerpentCS** [comment]
Travis fails!

File: library/models/library.py [hunk 19]
```diff
@@ -185,9 +185,9 @@ def onchange_day_to_return_book(self):
         @return : Dictionary having identifier of the record as key
                   and the book return date as value'''
         t = "%Y-%m-%d %H:%M:%S"
-        rd = relativedelta(days=self.day_to_return_book or 0.0)
-        if self.date_issue and rd:
-            ret_date = datetime.strptime(self.date_issue, t) + rd
+        diff = rd(days=self.day_to_return_book or 0.0)
```
Comment (JayVora-SerpentCS) [inline library/models/library.py:188]: is rd() a function or you made a shortcut?

File: school/models/parent.py [hunk 20]
```diff
@@ -67,13 +67,13 @@ def onchange_state(self, state_id):
         return {'value': {}}
 
 
-class ResUsers(models.Model):
+class ParentresUsers(models.Model):
```
Comment (JayVora-SerpentCS) [inline school/models/parent.py]: reason for the change of name?

File: library/models/library.py (hunk 19 above)
Comment (Arpit-SerpentCS) [reply library/models/library.py:188]: Here rd is short name given to relativedelta which is imported.  from dateutil.relativedelta import relativedelta as rd

File: school/demo/school_demo.xml [hunk 21]
```diff
@@ -492,62 +607,21 @@
...
         <!-- Student Informations -->
         <record id="demo_student_student_1" model="student.student">
             <field name="name">Mary</field>
             <field name="middle">Daniel</field>
             <field name="last">Richardson</field>
             <field name="state">draft</field>
             <field name="year" ref="demo_academic_year_1"/>
-            <field name="pid">2015/07</field>
+            <field name="pid">New</field>
             <field name="date_of_birth">03/04/2006</field>
```
Comment (JayVora-SerpentCS) [inline school/demo/school_demo.xml]: These dates still give errors!

**Pull request merged**